### DISCUSSION TITLE
Vehicle Constraints - Minor Issues, but otherwise working

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#53] [Show and tell]
![2023-11-20-wheeled-vehicle-constraint](https://github.com/jrouwe/JoltPhysics.js/assets/756488/4d5dfe91-059e-4e67-a106-0c004456a8a4)

I got a version of the Wheeled constraint working in my fork, but was running into issues getting the Vehicle Constraint to trigger its OnStep callback when wiring the IDL of the StepListeners into the Physics Engine. For now I instead exposed the OnStep method on the Constraint in my version of the fork and call it every frame, passing in the Delta and PhysicsEngine.

So far it is working for the vehicle, but it does seem to stick a bit to the walls when you get up close, given I think because it can't easily get the rotational torque along the car's Y-axis.

If I can get the other 2 vehicles (motorcycle and tank), and figure out the OnStep callback issue, it may work for another Example PR.

https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/vehicles

**Comment by PhoenixIllusion** [comment]
Main issue has been the OnStep method triggering a `null function or function signature mismatch`
![2023-11-20-wheeled-vehicle-onstep-listener-error](https://github.com/jrouwe/JoltPhysics.js/assets/756488/d3299ff8-e12b-49da-b9d5-2c0cd8d1cdb0)

**Reply by PhoenixIllusion** [reply]
It looks like it is calling Constraint::GetType for some reason after disabling the source-map and checking the raw wasm stack:
Expected:
`call_indirect (param i32 f32 i32)` -> `OnStep ( StepListener_ptr, float delta, PhysicsSystem_ptr)`
the indirect call can be looked up via the last stack items prior to call_indirect, and then passing the very last stack value to the indirect_table via Javascript side Instance.Table.get(indirect_i32_id) -> which resolves to $JPH::Constraint::GetType()

Possibly this is an issue with multiple inheritance? The call_indirect pointer is possibly jumping to one of the first post-destructor methods based on OnStepListener based on knowledge of that class struct, which on a pure StepListener this would be the OnStep method, but since it is also the VehicleConstraint class it has a much larger list of methods and it's jumping to GetType instead.

I could try writing a wrapper in the Jolt.h IDL file to create a dedicated OnStep listener containing a VehicleConstraint that only has the 1x method and forwards on to the VehicleConstraints OnStep, but the private visibility of the method makes things challenging.

**Reply by PhoenixIllusion** [reply]
Apparently just making a dedicated class wrapper, explicitly stating the typing was enough for the indirect_table to be setup and called correctly. It didn't matter that the method was private and had the multiple base class.
```
class VehicleConstraintStepListener: public PhysicsStepListener {
	public:
		VehicleConstraintStepListener(VehicleConstraint * inVehicleConstraint) {
			mInstance = inVehicleConstraint;
		}
		virtual void OnStep(float inDeltaTime, PhysicsSystem &inPhysicsSystem) override {
			PhysicsStepListener * instance = mInstance;
			instance->OnStep(inDeltaTime, inPhysicsSystem);
		}
	private:
		VehicleConstraint * mInstance;

};
```
Just need to make the motorcycle and tank demos now...

**Comment by PhoenixIllusion** [comment]
Functionally, main pain point is that if the car gets on the left or right side, the vehicle can't turn back towards the center while it's moving forward. I'm not yet sure how to resolve that
![2023-11-21-wheeled-vehicle-stuck-wall](https://github.com/jrouwe/JoltPhysics.js/assets/756488/e18b07e4-45cf-487b-9df5-f427423d24d7)

**Reply by PhoenixIllusion** [reply]
Cylinder Testers for the Wheels help some, but seem to prevent getting flush with the wall and tend to make the vehicle roll onto its side more often after hitting the wall. It may be needing more tweaking of the Cylinder Tester parameters

**Comment by PhoenixIllusion** [comment]
Motorcycle done... Tank remains, possibly some param testing/tweaking to confirm if the params are working as expected

![2023-11-21-motorcycle](https://github.com/jrouwe/JoltPhysics.js/assets/756488/6ffb8f0c-693f-4bff-bbbc-73fe61baa9ca)

**Comment by jrouwe** [comment]
This is very cool! Thanks for all the work, I'm looking forward to your PR!

> Cylinder Testers for the Wheels help some, but seem to prevent getting flush with the wall and tend to make the vehicle roll onto its side more often after hitting the wall. It may be needing more tweaking of the Cylinder Tester parameters

What may be causing the issue is that the wheels now stick out on both sides of the car. This means that they will intersect with the wall if the car body is flush against the wall and this can cause weird issues: Since the wheels are not really collision volumes but act as 'ray casts', the vehicle is not really pushed out of collision sideways. The vehicle will simply see this as collision that causes the suspension in the 'max up' position which means the springs have a lot of power and the vehicle will roll over. Another possible solution could be to add invisible boxes or trapezoid shaped convex hulls around the top of the wheels to prevent the intersection between wall and wheel:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/c257ba52-460f-4be9-bc75-b67ba8061220)

**Comment by PhoenixIllusion** [comment]
Took me a bit of panicking to figure out tanks need the treads engaged with either left/right Ratio in combo with Forward to move, so it just sat there and did nothing.

There is a chance I've been driving everything backwards since I kind of guessed what as forward until I got to Tank, which would be causing all prior demo's gears to never shift.  

I think the main change I want to do is standardize the variable setup between the 3 demos. The wheeled vehicle one uses the Unit test structure of an object with settings, and the other two use the Sample structure of Const at the top of the file.

![2023-11-21-tank](https://github.com/jrouwe/JoltPhysics.js/assets/756488/10c34d75-8d1e-4313-b714-dc385c63de72)

**Comment by PhoenixIllusion** [comment]
I think my wheel rotation seems to spin too much in some cases (just based on applying the `GetRotationAngle` along the rotation axis. It might just bee the friction causing slipping. At least the suspension working now and I got the vehicles facing what it considers 'forward' according to the steer angle being on front now.

![2023-11-22-tank-wheel-suspension-rotation](https://github.com/jrouwe/JoltPhysics.js/assets/756488/7364a27d-b59d-4d6a-b669-0b1ffe5f4a01)
![2023-11-22-car-wheel-rotation](https://github.com/jrouwe/JoltPhysics.js/assets/756488/c2564a03-5797-4010-b0f4-e9a538daea98)

**Reply by jrouwe** [reply]
Is there a reason why you cannot use ```VehicleConstraint.GetWheelWorldTransform```? Using this means you don't have to calculate the wheel rotation/steering vector yourself.

**Reply by PhoenixIllusion** [reply]
The WorldTransform could be used if you were going to wire the tires in a separate model instead of child-models, but since I had been storing them with the vehicle itself, they would more likely need to use the LocalTransform. The LocalTransform works if you add the wheels themselves to the model as direct children and do a pure copy of the local-transform to that child so it updates the Pos/Rotation/Scale. I'd found it educational to read the 3x floats instead of the read/parse/recreate/apply of the Mat44 every change.

My thoughts had also been potentially applying those values a non 1-to-1 setup as well, where those values (suspension, steer, and rotation) could be useful for non-exact animation of a model as well, possibly applied to a rig.

I think Mat44 in the JS library currently requires reading it out as 4x distinct Vec4's, with no trivial way to just do a float/double 16x MemCopy. The closest I've seen was marking the mF32 of the Vec4 as exposed in WebIDL as a VoidPtr and the doing a memory-read against the raw Heap location.

**Reply by jrouwe** [reply]
Ok, I'll first get all of this integrated and then I'll look at that later.
